fastlane/fastlane
# [spaceship] Better error handling of 403 status code when using API Key

### Checklist

### Motivation and Context


Resolves #18193

When using the API Key for App Store Connect API and the account have new program license agreements that needs to be signed, I was getting this error:
`undefined method 'status' for nil:NilClass`

Motivation for this PR is to fix this error and show a proper error message when there is new agreements needed to be signed.

### Description


In the client class we already handle the 403 status code, but the raise of the error class is causing the `status for nil:NilClass` error, when using the API Key.
Adding a check if client is kind of `PortalClient`, will prevent the status nil error. 

I then added handling of 403 status error to get a proper error message when new agreements needs to be signed.

### Testing Steps



I have not been able to trigger the current error handling of the 403 status code. 

To test the new error handling, you will need an API Key to an account where license agreements needs to be signed, so you are able to get this error response:
```json
{
    "errors": [
        {
            "id": "6LVKXRE2VZG23UUYXXXXXXXXXX",
            "status": "403",
            "code": "FORBIDDEN.REQUIRED_AGREEMENTS_MISSING_OR_EXPIRED",
            "title": "A required agreement is missing or has expired.",
            "detail": "This request requires an in-effect agreement that has not been signed or has expired.",
            "links": {
                "see": "/agreements"
            }
        }
    ]
}
```

Run with: `bundle exec rake debug --trace`
```ruby
task(:debug) do
  require 'spaceship'

  token = Spaceship::ConnectAPI::Token.from_json_file(File.expand_path("./authkey.json"))
  Spaceship::ConnectAPI.token = token

  begin
    app = Spaceship::ConnectAPI::App.find("dk.dinsen.cool.app")
    puts(app.to_yaml)
  rescue Spaceship::ProgramLicenseAgreementUpdated => ex
    puts(ex)
  end
end
```
